hcloud/client_handler_retry.go

```diff
@@ -30,13 +32,54 @@ func (h *retryHandler) Do(req *http.Request, v any) (resp *Response, err error)
 			// - request preparation
 			// - network connectivity
 			// - http status code (see [errorHandler])
-			if IsError(err, ErrorCodeConflict) {
-				time.Sleep(h.backoffFunc(retries))
-				retries++
-				continue
+			if ctx.Err() != nil {
+				return resp, ctx.Err()
+			}
+
+			if retryPolicy(resp, err) {
+				select {
+				case <-ctx.Done():
+					return resp, err
+				case <-time.After(h.backoffFunc(retries)):
+					retries++
+					continue
+				}
 			}
 		}
 
 		return resp, err
 	}
 }
+
+func retryPolicy(resp *Response, err error) bool {
+	if err != nil {
+		var apiErr Error
+
+		switch {
+		case errors.As(err, &apiErr):
+			if apiErr.Code == ErrorCodeConflict {
+				return true
+			}
+		case errors.Is(err, ErrStatusCode):
+			if resp == nil || resp.Response == nil {
+				// Should not happen
+				return false
+			}
+
+			if resp.Response.Request.Method != "GET" {
+				return false
+			}
+
+			switch resp.Response.StatusCode {
+			// 4xx errors
+			case http.StatusTooManyRequests:
+				return true
+			// 5xx errors
+			case http.StatusBadGateway, http.StatusServiceUnavailable, http.StatusGatewayTimeout:
+				return true
+			}
```

Comment:
`errorHandler.Do()` first tries to parse the body in a `hcloud.Error`. Only if thats not possible it returns an `ErrStatusCode`. So most of these will primarily result in `hcloud.Error` and not reach this code.

I would suggest also adding the related codes to `case errors.As(err, &apiErr)`:

- `ErrorCodeServiceError`
- `ErrorCodeRateLimitExceeded`
- `ErrorCodeServerError` (needs to be created for `"server_error"`)

Reply:
- `service_error` (HTTP 422) is used for very random scenarios, we must not retry on it, or we will have very weird errors.
- `rate_limit_exceeded` (HTTP 429) is now retried.
- `server_error` (HTTP 500) is a fallback error, so if the server cannot handle the error, we probably don't want to retry it.

I reorder the checks, if we have an api error code, if we have an http response, and if we hit a network timeout.